[shyanyong] iP

shyanyong opened the pull request:
# Brotherman
Brotherman helps you with tasks.
- text-based
-  easy to learn

To use brotherman:
1. download the code
2. idk run it 
3. add your tasks
4. have fun

this my `main`
```
    public static void main(String[] args) {
        new Brotherman("./data/brotherman.txt").run();
    }
```

Features:
- [ ] Manage tasks
- [ ] Manage Deadlines
- [ ] Manage Events :tada:

:shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: 

>  Someone **Cooked** Here

For a demo click [here](https://youtu.be/q6EoRBvdVPQ?si=whjVOxqt5tQBIHvw).

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

shyanyong committed:
Level 0. Rename, Greet, Exit
(a) Give your chatbot a new name, to differentiate it from the placeholder name Duke.

(b) Implement an initial skeletal version of the it that simply greets the user and exits.

shyanyong committed:
Level 1. Echo
Improve the skeletal version of Duke so that it echos commands entered by the user, and exits when the user types the command bye.

shyanyong committed:
Level 2. Add, List
Add the ability to store whatever text entered by the user and display them back to the user when requested.

shyanyong committed:
Level 3. Mark as Done
Add the ability to mark tasks as done. Optionally, add the ability to change the status back to not done.

shyanyong committed:
Level 4. ToDos, Events, Deadlines
Add support for tracking three types of tasks:

ToDos: tasks without any date/time attached to it e.g., visit new theme park
Deadlines: tasks that need to be done before a specific date/time e.g., submit report by 11/10/2019 5pm
Events: tasks that start at a specific date/time and ends at a specific date/time
e.g., (a) team project meeting 2/10/2019 2-4pm (b) orientation week 4/10/2019 to 11/10/2019

Added support for all the tasks

shyanyong committed:
Use the input/output redirection technique to semi-automate the testing of Duke.

shyanyong committed:
Updated automatated testing

shyanyong committed:
Level 5. Handle Errors
Teach the chatbot to deal with errors such as incorrect inputs entered by the user.

shyanyong committed:
Level 6. Delete
Add support for deleting tasks from the list.

shyanyong committed:
Updated tests

shyanyong committed:
Level-7: Implement automatic task data saving and loading from disk
This commit introduces the feature of automatically saving tasks to the hard disk whenever the task list undergoes changes. Additionally, it enables the loading of task data from the hard disk during the chatbot's startup. The file name and relative path are hard-coded from the project root. The data is stored in a format designed for efficient storage and retrieval. This feature enhances the user experience by ensuring that task data persists across chatbot sessions and remains consistent even in the event of interruptions.

shyanyong committed:
Merge branch 'branch-Level-7'

shyanyong committed:
Level-8: Dates and Times
This commit implements date and time understanding for tasks.

This commit enhances the chatbot's capability to intelligently interpret dates and times provided in commands. When processing task inputs like "deadline return book /by 2/12/2019 1800", the chatbot now comprehends the date "2/12/2019" as 2nd of December 2019 and the time "1800" as 6pm. The parsed dates are stored as java.time.LocalDate or java.time.LocalDateTime objects in task instances.

Additionally, the chatbot accepts dates in the yyyy-mm-dd format (e.g., 2019-10-15) and presents them in a more readable format such as MMM dd yyyy (e.g., Oct 15 2019) when displaying task details. This improvement elevates the user experience by enabling more natural and intuitive interaction with date and time information.

shyanyong committed:
Merge branch 'branch-Level-8'

shyanyong committed:
A-MoreOOP:
Rewrote brotherman chatbot in OOP style.  Added new classes to encapsulate commands as well as Ui, Storage and TaskList.

shyanyong committed:
Merge branch 'branch-MoreOOP'

Merge MoreOOP.  OOP the files.  Created new classes to encapsulate commands, parsing, ui and storage.

shyanyong committed:
A-Packages:

Added packages into Brotherman.  Added Commands, exceptions, parser, storage, tasks, ui.

shyanyong committed:
A-Packages:

Added packages into Brotherman.  Added Commands, exceptions, parser, storage, tasks, ui.

shyanyong committed:
Merge branch 'branch-Packages'

shyanyong committed:
Merge branch 'add-gradle-support' into branch-Gradle

shyanyong committed:
Merge branch 'branch-Gradle'

shyanyong committed:
Add JUnit tests for Parser class and Ui class.

Sure, I can help you write a GitHub commit message where JUnit tests have been added to test the code. Here is an example of a commit message:

Add JUnit tests for Parser class

This commit adds JUnit tests for the parseTodo() and ShowWelcomeMessage()

shyanyong committed:
Add Javadocs to all files

This commit adds Javadocs to all files in the project. Javadocs are documentation comments that can be used to describe the purpose of a class, method, or field. They can be helpful for understanding the code and for debugging.

shyanyong committed:
Add FindCommand class

This commit adds the FindCommand class, which allows users to find tasks with a specified keyword.

The FindCommand class is a subclass of the Command class. It has a constructor that takes the keyword as a parameter. The execute() method of the FindCommand class calls the getTasksByKeyword() method of the TaskList class to get the tasks that match the keyword. The showTaskList() method of the Ui class is then called to display the list of tasks to the user.

shyanyong committed:
Merge branch 'branch-A-JavaDoc'

shyanyong committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/brotherman/tasks/TaskList.java
#	src/main/java/brotherman/ui/Ui.java

JasonLCY-Temp commented on src/main/java/Brotherman.java:
```diff
@@ -0,0 +1,53 @@
+/**
+ * Represents a Brotherman chatbot
+ */
+public class Brotherman {
```
Is this class intended to not be in a package?

JasonLCY-Temp commented on src/main/java/brotherman/commands/FindCommand.java:
```diff
@@ -0,0 +1,19 @@
+import brotherman.storage.Storage;
+import brotherman.tasks.TaskList;
+import brotherman.ui.Ui;
+public class FindCommand extends Command {
+
+    private String keyword;
+
+    public FindCommand(String keyword) {
+        super(false);
+        this.keyword = keyword;
+    }
```
Considering you have basically 100% class and method javadoc coverage everywhere else (great effort on your part), are the missing javadocs in this class intended?

nknguyenhc commented on build.gradle:
```diff
@@ -0,0 +1,43 @@
+}
+
+application {
+    mainClass.set("seedu.duke.Duke")
```
Class name does not appear to be correct, gradle therefore is unable to run your project.

nknguyenhc commented on src/main/java/brotherman/parser/Parser.java:
```diff
@@ -0,0 +1,212 @@
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
```
Import statements are in the wrong order. Imports from `java` should be at the top, followed by custom packages (`brotherman` packages). Furthermore, please make sure that imports follow lexicographical order within each group.

nknguyenhc commented on src/main/java/brotherman/storage/Storage.java:
```diff
@@ -0,0 +1,87 @@
+     */
+    public Storage(String filePath) {
+        this.folder = new File(filePath);
+        if(!folder.exists()) folder.mkdirs();
```
There should be a white space following the `if` keyword, and there should be a curly bracket block even if there is only one statement. Please follow the Java coding standard [here](https://se-education.org/guides/conventions/java/intermediate.html).

nknguyenhc commented on data/brotherman.txt:
```diff
@@ -0,0 +1,3 @@
+D|true|read book/by03/03/1234
+E|false|eat ass/from01/01/1234/to12/12/9999
+D|false|eat/by01/01/1221
```
Files generated by your code should not be committed, instead should be added to `.gitignore`.

nknguyenhc commented on src/main/java/brotherman/tasks/Event.java:
```diff
@@ -0,0 +1,81 @@
+     */
+    @Override
+    public String storeText() {
+        return String.format("%s|%s|%s/from%s/to%s", this.type(), this.isDone, this.description, getStartTime(), getEndDateTime());
```
Line is too long. Please break line to make your code more readable.

nknguyenhc commented on src/main/java/brotherman/tasks/TaskList.java:
```diff
@@ -0,0 +1,103 @@
+
+    /**
+     * Returns the task at the specified index
+     * @param index Index of the task
```
Unused Javadoc tag `@param index`. Please make sure that once you changed the method API, you change the corresponding Javadoc comment as well.

nknguyenhc commented on src/main/java/brotherman/ui/Ui.java:
```diff
@@ -0,0 +1,99 @@
+
+import java.util.ArrayList;
+import java.util.Scanner;
+import brotherman.tasks.*;
```
Please do not use `*` import. Name the classes you use from the import, so that classes used in your code can be traced to the imports.

nknguyenhc commented on src/main/java/Brotherman.java:
```diff
@@ -0,0 +1,53 @@
+                command.execute(taskList, ui, storage);
+                isExit = command.isExit();
+            } catch (BrothermanException e) {
+                System.out.println("should be a show error message");
```
Error message can be improved to feedback to the user of the error in the command.

nknguyenhc commented on src/main/java/brotherman/commands/FindCommand.java:
```diff
@@ -0,0 +1,19 @@
+
+    @Override
+    public void execute(TaskList task, Ui ui, Storage storage) {
+        ui.showFind(task, keyword);
```
`showFind` is not found in `Ui` class. This might be due to merge conflicts from your latest merge. Please make sure that after you merge, before you make a commit, the code is still functional.

nknguyenhc commented on src/main/java/brotherman/commands/Command.java:
```diff
@@ -0,0 +1,42 @@
+     * @param storage Storage to save the task list
+     */
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+
```
This should be declared as an abstract method instead of a method without implementation. This is so that you can make sure all child classes override this method.

shyanyong committed:
Add Checkstyle to project and resolve all Checkstyle issues

In this commit, we introduce Checkstyle to our project to enforce consistent coding style and maintain code quality. The following changes have been made:

- Added Checkstyle plugin to IntelliJ IDEA for integrated code analysis.
- Configured Checkstyle with a custom checkstyle.xml configuration file tailored to our project's coding standards.
- Resolved all Checkstyle issues identified in the codebase, ensuring adherence to the coding guidelines.

This implementation helps us maintain a cleaner and more consistent codebase, making it easier to collaborate and maintain the project in the long run.

shyanyong committed:
Merge branch 'branch-A-Checkstyle'

nknguyenhc commented on src/main/java/brotherman/parser/Parser.java:
```diff
@@ -0,0 +1,212 @@
+            String endTime = parts[2];
+
+            return new AddCommand(parseEvent(desc, startTime, endTime));
+        } else {
```
A bunch of `if` - `else if` - `else` keywords here. You may consider using a `switch` statement.

nknguyenhc commented on src/main/java/brotherman/parser/Parser.java:
```diff
@@ -0,0 +1,212 @@
+        try {
+            startDateTimeObj = LocalDateTime.parse(startTime, formatter);
+            endDateTimeObj = LocalDateTime.parse(endTime, formatter);
+        } catch (DateTimeParseException e) {
```
Exceptions should not be used as part of logic flow.

nknguyenhc commented on src/main/java/brotherman/parser/Parser.java:
```diff
@@ -0,0 +1,212 @@
+            return new ExitCommand();
+        }
+
+        isValidCommands(userCommand);
```
Where does this method come from? Please make sure to indicate where the method come from instead of leaving it to the compiler to infer, so that your code is more traceable.

nknguyenhc commented on src/main/java/brotherman/parser/Parser.java:
```diff
@@ -0,0 +1,212 @@
+     * @return True if the user input is valid
+     * @throws BrothermanException If the user input is invalid
+     */
+    public static boolean isValidCommands(String command) throws BrothermanException {
```
This method both throws an exception and returns a boolean. However, the boolean returned is always `true` if the method does not throw an error. By the name of this method, maybe it is better for this method to return a boolean and never throw an exception?

nknguyenhc commented on src/main/java/brotherman/parser/Parser.java:
```diff
@@ -0,0 +1,212 @@
+                LocalDate endDate = LocalDate.parse(endTime, formatter);
+                startDateTimeObj = startDate.atStartOfDay();
+                endDateTimeObj = endDate.atStartOfDay();
+            } catch (DateTimeParseException ex) {
```
Im not sure if it is appropriate to handle the exception here. If there is an exception, the programme will continue on with a `null` value returned from this programme. This can cause `NullPointerException` in your programme even when the parse error is already handled here.

nknguyenhc reviewed (commented):
Generally, there are quite a few coding standard violations. After you have completed `A-CodingStandard` level, please make sure that your code follows coding standard by consistently running the style checks.

Code qualities can be improved here and there. Please make sure that your programme is functional before every commit, especially after a merge with conflicts. Also, please make sure that you handle exceptions at the correct place, so that your programme does not crash.

shyanyong committed:
Resolve Parsing and general Code Cleanup

   - Improved parsing logic for date and time strings to handle a wider range of formats.

- Conducted a thorough code review and performed extensive code cleanup.
   - Removed redundant code, optimized data structures, and enhanced code readability.
   - Applied consistent coding standards and naming conventions across the project.

shyanyong committed:
Add simple GUI

Implemented a basic graphical user interface (GUI) using JavaFX. This GUI includes a main window with a scrollable chat container, a user input field, and a send button. The GUI provides the foundation for future user interaction and chatbot functionality.  Logic for chatbot not yet done.

shyanyong committed:
Enhance GUI using JavaFX and maintain chatbot functionality
In this commit, I have improved the graphical user interface (GUI) of our application by integrating JavaFX components. The new GUI not only enhances the overall user experience but also provides a more modern and intuitive look and feel.
Additionally, we have updated the application's logic to seamlessly integrate the JavaFX components while ensuring that the core functionality of the chatbot remains intact. This means that users can continue to interact with the chatbot as they did before, with all existing features and capabilities fully preserved.

shyanyong committed:
Merge branch 'Branch-Level-10'

shyanyong committed:
Add assertions

Added assertions in some key places.

Ensures that important assumptions hold at various points in the code.

shyanyong committed:
Improve Code Quality

Improved the overall general code quality of the project

shyanyong committed:
Merge pull request #1 from shyanyong/branch-Assertions

Add assertions

shyanyong committed:
Merge pull request #2 from shyanyong/branch-CodeQuality

Improve Code Quality

shyanyong committed:
Add Help function

Provide in-App guidance to users.

Added command to help page for users to view commands.

shyanyong committed:
Merge branch 'C-Help'

shyanyong committed:
Merge branch 'master' of https://github.com/shyanyong/ip

# Conflicts:
#	src/main/java/brotherman/ui/Ui.java

shyanyong committed:
refactor: Cleaned up code, introduced enums, and added switch cases

In this commit, I have performed several code improvements to enhance readability and maintainability. The main changes include:

- Cleaned up the codebase by removing redundant and unused variables, functions, and comments.
- Introduced meaningful enums to replace magic numbers or string literals, improving code clarity and reducing the chance of errors.
- Added switch cases to handle different scenarios more efficiently and comprehensively.

These changes will make the codebase more organized and easier to understand, and they also lay the groundwork for future enhancements. The code now adheres to better coding standards and practices.

shyanyong committed:
Refactor code for improved clarity and maintainability

In this commit, I've cleaned up the codebase to enhance readability and maintainability. I applied the SLAP (Single Level of Abstraction Principle) to ensure that each function focuses on a single task, resulting in cleaner and more modular code. Additionally, I've aimed to reduce the length of functions to less than 30 lines wherever possible.

Changes made in this commit:
- Split complex functions into smaller, more focused ones.
- Renamed variables and functions to improve code clarity.
- Removed redundant comments and code.
- Ensured consistent code formatting and style.

These changes should make the codebase easier to understand, maintain, and extend in the future.

shyanyong committed:
Refine GUI: Add Colored Background and Rounded Images

In this commit, the GUI has been enhanced to provide a more visually appealing
user experience. The changes include:

- Added colored background for the chat messages to improve readability and
  visual separation between user and bot messages.

- Rounded the user and bot images to create a more polished and modern look,
  enhancing the overall aesthetic of the chat interface.

These changes improve the overall look and feel of the Brotherman chatbot, making
it more user-friendly and visually appealing.

#GUI #Design #UserExperience

shyanyong committed:
Merge branch 'branch-BetterGui'

shyanyong committed:
update readme

shyanyong committed:
updated readme

shyanyong committed:
Fix Bugs.

Fixed storage bugs.  Recreated Jar file
